packages/components/src/number-control/test/index.tsx

```diff
@@ -0,0 +1,616 @@
+								.valid
+						)
+					}
+				/>
+			);
+
+			const input = getInput();
+
+			await user.clear( input );
+			await user.type( input, '14' );
+
+			expect( input ).toHaveValue( 14 );
+
+			await user.keyboard( '[Enter]' );
+
+			expect( input ).toHaveValue( 10 );
+
+			expect( onChangeSpy ).toHaveBeenCalledTimes( 4 );
+
+			// First call: clear value
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 1, '', true );
+			// Second call: invalid, unclamped value
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 2, '1', true );
+			// Third call: invalid, unclamped value
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 3, '14', false );
+			// Fourth call: valid, clamped value
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 4, 10, true );
+		} );
+	} );
+
+	describe( 'Validation', () => {
+		it( 'should clamp value within range on ENTER keypress', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } min={ 0 } max={ 10 } /> );
+
+			const input = getInput();
+
+			await user.clear( input );
+			await user.type( input, '-100' );
+			await user.keyboard( '[Enter]' );
+
+			/**
+			 * This is zero because the value has been adjusted to
+			 * respect the min/max range of the input.
+			 */
+			expect( input ).toHaveValue( 0 );
+		} );
+
+		it( 'should clamp value within range on blur', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } min={ 0 } max={ 10 } /> );
+
+			const input = getInput();
+			await user.clear( input );
+			await user.type( input, '41' );
+
+			// Before blurring, the value is still un-clamped
+			expect( input ).toHaveValue( 41 );
+
+			// Blur the input
+			await user.click( document.body );
+
+			// After blur, value is clamped
+			expect( input ).toHaveValue( 10 );
+		} );
+
+		it( 'should parse to number value on ENTER keypress when required', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } required={ true } /> );
+
+			const input = getInput();
+			await user.clear( input );
+			await user.type( input, 'abc' );
+			await user.keyboard( '[Enter]' );
+
+			expect( input ).toHaveValue( 0 );
+		} );
+
+		it( 'should parse to empty string on ENTER keypress when not required', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } required={ false } /> );
+
+			const input = getInput();
+			await user.clear( input );
+			await user.type( input, 'abc' );
+			await user.keyboard( '[Enter]' );
+
+			// React Testing Library associated `null` to empty string types for
+			// numeric `input` elements
+			// (see https://github.com/testing-library/jest-dom/blob/v5.16.5/src/utils.js#L191-L192)
+			expect( input ).toHaveValue( null );
```

Comment:
Previously, in these two tests we were using `fireEvent` to submit a new value `10 abc` — the fact that the value was all changed in one go meant that the whole string couldn't be parsed as a number.

Switching to `user-event` meant that each single character in that string would be typed individually. This change was causing the tests to fail, because the component would first parse the number `10` correctly, and then refuse to parse `' abc'` — meaning that the input's value would become `10` in both tests, instead of `0` and `null`.

I therefore decided to change slightly the tests, and simulate the user entering a fully invalid number (ie. `'abc'`), which makes the tests pass again.

Reply:
A question related to the second of these tests: Do we expect the visible value in regular browsers to actually clear? That's what the test suggests to me but it's not the case.

https://user-images.githubusercontent.com/9000376/203652638-c19148e0-67c1-46f2-8191-efba05a43492.mp4

If that's what we expect, I don't think it can actually be tested by jsdom or even very easily in most browsers because in number inputs, non-numeric values are reported as ''. That means the second test is really only good for testing that `NumberControl` doesn't produce a numeric value in this scenario. I suppose there’s some value in that but maybe the test could be more clearly written that way.

I don't think it’s anything this PR should address. I just wanted to bring it up.

Reply:
Good point! I tried to make some small improvements in https://github.com/WordPress/gutenberg/pull/45990/commits/6732efb21b608cad27ba5fa3d243c9899cdeb907, hopefully it's a step in the right direction!